File: 1962-remove-stones-to-minimize-the-total/1962-remove-stones-to-minimize-the-total.cpp
/**
* Leetcode Problem - https://leetcode.com/problems/remove-stones-to-minimize-the-total/
* Author - Shubham Nagaria
* Date - 22nd April 2022
Remove Stones to Minimize the Total
**/
class Solution {
public:
    int minStoneSum(vector<int>& piles, int k) {
        int n = piles.size();
        priority_queue<int>pq;
        long int sumi =0;
        for(int i: piles){
            sumi += i;
            pq.push(i);
        }
        while(k--){
           int x= pq.top();
            pq.pop();
            if(x==0)
                break;
            sumi -= x;
            x-=floor(x/2);
            sumi +=x;
            pq.push(x);
        }
        return sumi;
    }
};

File: 1962-remove-stones-to-minimize-the-total/1962-remove-stones-to-minimize-the-total.py
# Date - 24th April 2022
import heapq
class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        sumi =sum(piles)
        piles=[-i for i in piles]
        heapq.heapify(piles)
        while k>0:
            k-=1
            maxi = -heapq.heappop(piles)
            sumi-=maxi
            maxi -= maxi//2
            sumi +=maxi
            heapq.heappush(piles,-maxi)
        return sumi